FATES LUH2 data curation tool

glemieux opened the pull request:
### Description:
This pull request adds the initial version of the fates luh2 data curation pipeline.  The tool consists of four parts: 1) a module, luh2mod.py, that condenses a majority of the luh2 updates into a set of functions, 2) a python script, luh2.py, that calls the module functions and takes a set of command line arguments, 3) a bash script, luh2.sh, that sets up the expected arguments and calls the python script and 4) a conda environment `yml` file to provide the user with the necessary dependencies to run the tool.

This partially addresses https://github.com/NGEET/fates/issues/450.  The script was heavily adapted from work done by @ckoven.

Raw LUH2 data from https://luh.umd.edu/ is available on cheyenne in the input data directory under `/lnd/clm2/rawdata/LUH2`.

### Collaborators:
@ckoven @adrifoster 

### Expectation of Answer Changes:
No changes as this only impacts tooling

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [ ] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [ ] FATES PASS/FAIL regression tests were run
- [ ] If answers were expected to change, evaluation was performed and provided


### Test Results:
Testing not necessary

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

<!--- paste in test results here -->


<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

glemieux committed:
Add luh2 module and conda environment yaml file

glemieux committed:
Add function to update luh2 time units

glemieux committed:
Update conda environment for luh2 tools

glemieux committed:
Adding functionality notes

glemieux committed:
Start bringing in Charlie's xesmf code

glemieux committed:
Update yml environment file

glemieux committed:
Continue converting Charlie's script

glemieux committed:
Update notes

glemieux committed:
Update comments

glemieux committed:
Renamed luh2 module and added primary script

glemieux committed:
Moving script into module function

glemieux committed:
Intermediate improvements

glemieux committed:
Fixup

glemieux committed:
Update to xarray 0.19.0

This incorporates an update to handle a pandas deprication issue when
using cftime objects.  See https://github.com/pydata/xarray/pull/5359

glemieux committed:
Truncate to the yml file and update prep function

Having too many of the dependencies in the yml file was causing issue
with the dependency order and limiting xesmf to an older version.

glemieux committed:
more intermediate updates

glemieux committed:
Add command line argument parsing

glemieux committed:
continuing to add updates and debug

glemieux committed:
add year variable as copy of time dimension

This is necessary due to a ncd_io read requirement set in dynsubgrid.

glemieux committed:
converting variables to snake case

glemieux committed:
adjust regridding methods

glemieux committed:
various fixes and improvements

glemieux committed:
Separated luh2 module and luh2 call script

glemieux committed:
More updates

glemieux committed:
start working on allowing read in of regridder file

glemieux committed:
Fix regridding loop

glemieux committed:
Remove unnecessary xarray import call

glemieux committed:
Adding state correction function

glemieux committed:
Add handling of regridder argument option

glemieux committed:
switch the importdata and prepdataset functions

glemieux committed:
Correct some issues

glemieux committed:
adding type checks

glemieux committed:
adding check against the time variable for the input dataset

glemieux committed:
fix import function

glemieux committed:
fix argument call

glemieux committed:
update check function outputs

glemieux committed:
Add bash script

glemieux committed:
refactoring luh2 module

This refactor is to account for a misunderstanding in the way that
the reuse of a saved regridder weights file is to be applied.

glemieux committed:
Add merge flag to import data

This avoids prepping the data if its already been prepped

glemieux committed:
fix order of operations

glemieux committed:
add commands to remove intermediate file copies

This helps mitigate loss of storage space

glemieux committed:
move all luh2 code into named folder

rosiealice commented:
This is awesome Greg and Charlie. Thank you :)

glemieux committed:
convert YEAR from cftime object to number

glemieux commented:
I'm currently struggling with an issue that seems to have come up out of nowhere for some reason.  The regridding loop seems to hit a runtime error.  Trawling through various issues and stack exchange, the best I can guess is this is an issue with xarray, but I'm not entirely confident:

```
Traceback (most recent call last):
  File "/home/glemieux/Repos/fates/tools/luh2/luh2.py", line 127, in <module>
    main()
  File "/home/glemieux/Repos/fates/tools/luh2/luh2.py", line 47, in main
    regrid_luh2,regridder_luh2 = RegridConservative(ds_luh2, ds_regrid_target,
  File "/home/glemieux/Repos/fates/tools/luh2/luh2mod.py", line 217, in RegridConservative
    ds_regrid = RegridLoop(ds_to_regrid, regridder)
  File "/home/glemieux/Repos/fates/tools/luh2/luh2mod.py", line 265, in RegridLoop
    ds_regrid[ds_varnames[i]] = regridder(ds_to_regrid[ds_varnames[i]])
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xesmf/frontend.py", line 456, in __call__
    return self.regrid_dataarray(
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xesmf/frontend.py", line 536, in regrid_dataarray
    dr_out = xr.apply_ufunc(
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/computation.py", line 1205, in apply_ufunc
    return apply_dataarray_vfunc(
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/computation.py", line 312, in apply_dataarray_vfunc
    result_var = func(*data_vars)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/computation.py", line 679, in apply_variable_ufunc
    input_data = [
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/computation.py", line 680, in <listcomp>
    broadcast_compat_data(arg, broadcast_dims, core_dims)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/computation.py", line 595, in broadcast_compat_data
    data = variable.data
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/variable.py", line 430, in data
    return self.values
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/variable.py", line 603, in values
    return _as_array_or_item(self._data)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/variable.py", line 309, in _as_array_or_item
    data = np.asarray(data)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/indexing.py", line 664, in __array__
    self._ensure_cached()
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/indexing.py", line 661, in _ensure_cached
    self.array = NumpyIndexingAdapter(np.asarray(self.array))
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/indexing.py", line 634, in __array__
    return np.asarray(self.array, dtype=dtype)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/indexing.py", line 535, in __array__
    return np.asarray(array[self.key], dtype=None)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/coding/variables.py", line 73, in __array__
    return self.func(self.array)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/coding/variables.py", line 145, in _apply_mask
    data = np.asarray(data, dtype=dtype)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/indexing.py", line 535, in __array__
    return np.asarray(array[self.key], dtype=None)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/backends/netCDF4_.py", line 91, in __getitem__
    return indexing.explicit_indexing_adapter(
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/core/indexing.py", line 826, in explicit_indexing_adapter
    result = raw_indexing_method(raw_key.tuple)
  File "/home/glemieux/local/conda/miniconda3/envs/luh2/lib/python3.10/site-packages/xarray/backends/netCDF4_.py", line 104, in _getitem
    array = getitem(original_array, key)
  File "src/netCDF4/_netCDF4.pyx", line 4940, in netCDF4._netCDF4.Variable.__getitem__
  File "src/netCDF4/_netCDF4.pyx", line 5898, in netCDF4._netCDF4.Variable._get
  File "src/netCDF4/_netCDF4.pyx", line 2014, in netCDF4._netCDF4._ensure_nc_success
RuntimeError: NetCDF: HDF error
```

ckoven committed:
bugfixes to reduce memory usage and interpret cftime reference point

glemieux committed:
Merge pull request #28 from ckoven/fates-luh2_data

bugfixes to reduce memory usage and interpret cftime reference point

ekluzek commented:
I'll let @slevis-lmwg also answer. But, I also chatted with @wwieder and @ckoven about this and have a picture of what I think should happen.

We'll add the creation of these luh_timeseries files needed for FATES to the Makefile.data that's in tools/mksurfdata_esmf that creates the CTSM5.2 datasets. That makefile creates all the surface datasets, landuse timeseries datasets, using mksurfdata_esmf as well as single point datasets using surbset_data. So the usage here will be similar to that of subset_data in the mksurfdata_esmf Makefile.

If ctsm5.2 happens before this PR is ready we'll add the run of the luh_timeseries datasets to the Makefile in mksurfdata_esmf. If this PR is ready before that time -- we'll have to decide what to do, but my inclination would be to leave it like it is, and add the run of these datasets to the mksurfdata_esmf makefile later.

slevis-lmwg commented:
@ckoven 
@ekluzek's assessment makes sense to me.
I would be happy to review this PR. For greatest efficiency (at least for me), I would request that we go over the PR together (me, you, Greg, Erik?) in a meeting. Thanks!

glemieux commented:
> @ckoven @ekluzek's assessment makes sense to me. I would be happy to review this PR. For greatest efficiency (at least for me), I would request that we go over the PR together (me, you, Greg, Erik?) in a meeting. Thanks!

Having a meeting to review this sounds good to me.  Is there going to be a ctsm software meeting tomorrow?  Maybe we can figure out a time then?

slevis-lmwg commented:
Yes, we decided that we would meet tomorrow, and I plan to attend.

glemieux commented:
Per discussion with @slevis-lmwg today, it seems like getting a call to this toolchain into the makefile for ctsm5.2 should be an achievable goal.  This PR still needs testing on cheyenne (and other machines), which is the next step.  Right now, one of the main unknowns that we came out of from the meeting is can the python package`xesmf` be added to the necessary ctsm python environment?  Note that this utilizes `esmpy` on the backend for computing the regridding weights, [but uses `scipy` for applying them](https://xesmf.readthedocs.io/en/latest/notebooks/Backend.html#xESMF-backend-usage-and-benchmark).  

@ekluzek what do you think about adding `xesmf` to the `ctsm_pylib` environment?  If I recall, adding more dependencies is not preferred, in that it potentially adds more conflicts to deal with.   Alternatively, I noticed that `esmpy` is part of the `ctsm_pylib` currenlty; maybe I could refactor this code to utilize `esmpy` instead?

slevis-lmwg commented:
@glemieux we discussed some of these issues in today's software eng. meeting. Please look at the notes taken in the meeting's google doc.

glemieux commented:
[ ] Test on cheyenne

glemieux committed:
implement decode_times=False

ckoven commented:
my ncks version, which works, is:

``ncks --version``
``NCO netCDF Operators version 5.1.6 "Crossroads" built by nobody on Mac-1683965091281.local at May 13 2023 09:43:39
ncks version 5.1.6``

ckoven commented:
I put the states.nc file on Cheyenne at `/glade/scratch/charlie/states.nc`

running `ncks -v secmb states.nc -o test.nc` on cheyenne using that file worked for me.

testing the ncks version that I get on Cheyenne when I do a `module load nco` gives the following:

``ncks --version``
``NCO netCDF Operators version 5.1.4 "Bakhmut" built by csgteam on cheyenne3 at Feb 21 2023 15:42:02
ncks version 5.1.4``

glemieux commented:
Looks like this was indeed a corrupt file issue.  The states.nc file that I initially tested on cheyenne was a copy of the one from lobata.  Using a copy of the file Charlie placed on cheyenne, `ncks -v secmb` ran successfully for me as well.  I was also able to successfully run the luh2.py call on cheyenne as well.

glemieux committed:
remove call to attribute update in shell script

glemieux committed:
reworking luh2 shell script call

glemieux commented:
> * [x]  Test on cheyenne

I was able to successfully run the luh2.sh script on cheyenne using a custom conda environment with `xesmf` packaged installed based on the ["latest" ctsm conda env create configuration file](https://github.com/ESCOMP/CTSM/blob/4a868c1ccd700e78b5a2c2f064441a1d612c57bf/python/conda_env_ctsm_py_latest.txt).  

@slevis-lmwg I think this is good to review now.  I'm assuming we'll need to work out what update to the "standard" conda environment will need to be on the ctsm5.2 development branch.

glemieux committed:
fix choice range and add shell script input options

slevis-lmwg commented on tools/luh2/luh2.sh:
```diff
@@ -0,0 +1,57 @@
+OUTPUT_LOC=$4
+STATES_FILE=states.nc
+TRANSITIONS_FILE=transitions.nc
+MANAGE_FILE=management.nc
```
@glemieux and I discussed the risk of using the same three file names (states, transitions, management) for all cases that are present in /glade/p/cesmdata/cseg/inputdata/lnd/clm2/rawdata/LUH2. The most likely risk that comes to mind is one where we copy or move files without keeping track of the contents. We looked in a few files for metadata that may help differentiate between files but did not find enough information.

@glemieux and I recommend renaming the files according to case/scenario/version/other identifying characteristics. Doing so requires modifications to the tool, but maybe just in these few lines?

I suggested that @glemieux move forward without changing the file names for now. When the tool gets added to Erik's mksurfdata_esmf makefile, we can determine whether changing the file names adds significant complication. We may also decide to keep the file names unchanged in order to maintain "purity" relative to the original source of the data on the LUH2 website.

glemieux replied:
Converted the comment to #1058 for future discussion.

slevis-lmwg commented on tools/luh2/luh2.py:
```diff
@@ -0,0 +1,128 @@
+    print("generating output: {}".format(output_file))
+    regrid_luh2.to_netcdf(output_file)
+
+def CommandLineArgs():
```
In the add_argument calls, I mentioned the `default` option as possibly helpful.

I showed an example in this file (from the ctsm5.2 branch):
/glade/work/slevis/git/mksurfdata_toolchain/tools/mksurfdata_esmf/gen_mksurfdata_namelist.py

glemieux replied:
resolved via 8cbb7ab204c2f2eb397d2cfec1da8f3471506da6

slevis-lmwg commented on tools/luh2/luh2mod.py:
```diff
@@ -0,0 +1,258 @@
+#             SetMaskSurfData(input_dataset)
+#         print("mask added")
+#
+#     return(input_dataset)
```
@glemieux 
reminder to delete obsolete commented out code here and elsewhere.

glemieux replied:
completed with e44bc90ffb895100051b7ed6305a4fa3625c5db9

slevis-lmwg commented on tools/luh2/luh2mod.py:
```diff
@@ -0,0 +1,258 @@
+        dsflag = True
+    else:
+        dstype = 'Unknown'
+        print("Unrecognize data set")
```
@glemieux 
reminder to exit here with error message.

glemieux replied:
resolved with 4615b8a6a16462c6712d81939a2e9e3c89eff50c

slevis-lmwg reviewed (approved):
@glemieux feels this PR is ready to merge, after some final clean-up. Thank you @glemieux for going over the PR with me.

I am approving with a few minor comments to be addressed:
- Most important is the addition of comments + README for users unfamiliar with the tool
- Clean-up of obsolete commented out code
- One or two other minor things

Otherwise, as written the tool is easy to follow, and I do not feel expert enough in Python to recommend any type of refactor.

glemieux committed:
Add comments and remove unnecessary dependence to yaml file

glemieux committed:
add luh2 readme

glemieux commented on tools/luh2/conda-luh2.yml:
Completed with 63124fd1ef0422595e09b46a2b3bf759166d509e

glemieux committed:
remove old comments

glemieux committed:
Add system exit if unrecognized file type provided as argument

glemieux committed:
update luh2 tool argparse options and time handling

Adding 'default' and 'required' option as necessary to the argparse.
This helps remove some of the luh2mod code that was previously
handling default options.

This also adds code to handle non-historical luh2 data file input
arguments.  Prior to this, it was assumed that only historical
data files starting at 0850-01-01 were to be used.  Now this
can handle luh2 scenario files.  It also adds time bounds error handling.

rgknox commented on tools/luh2/README.md:
```diff
@@ -0,0 +1,55 @@
+# FATES LUH2 data tool README
+
```
Nice description here in the README, thanks

rgknox commented on tools/luh2/luh2.py:
```diff
@@ -0,0 +1,137 @@
+
+# LUH2 python script
+# Usage: python luh2.py -l <luh2file> -s <luh2staticfile>
+
```
Usage should match the README, right?
python luh2.py -l <raw-luh2-datafile> -s <luh2-static-datafile> -r <regrid-targetfile> -w <regridder-output> -o <outputfile>

glemieux replied:
Good catch.  Resolved per  9cbfe4f0a10196e7805aec032205746b455eb92b

glemieux committed:
update the usage description in luh2.py

glemieux commented:
> > * [x]   Test on cheyenne
> 
> I was able to successfully run the luh2.sh script on cheyenne using a custom conda environment with `xesmf` packaged installed based on the ["latest" ctsm conda env create configuration file](https://github.com/ESCOMP/CTSM/blob/4a868c1ccd700e78b5a2c2f064441a1d612c57bf/python/conda_env_ctsm_py_latest.txt).

I didn't make this very clear in this comment, but I drafted a pull request on the ctsm side to capture the custom conda environment: https://github.com/ESCOMP/CTSM/pull/2075

Per an google chat discussion with Ryan, I think we should address #1033 in the near future with another PR to make sure we can guarantee this and other fates python tools work for a particular set of package versions.

glemieux commented:
> Per an google chat discussion with Ryan, I think we should address #1033 in the near future with another PR to make sure we can guarantee this and other fates python tools work for a particular set of package versions.

I've drafted #1059 to partially address this.
